### PR TITLE
Address scattergl performance issue with TypeArrays

### PR DESCRIPTION
**Pull request by Seranicio**
Address: #4401

Overview of the Pull Request:

Address an unavoidable loop phase inside color clean function where
it iterates all over the graph points if data used is a derivative of a ArrayBuffer.
This causes graph to have way less performance relative to Array types.

Also added a little performance enhancement to calc function. Tell if this is out of scope :)

Simple Demo: [here](https://codepen.io/Seranicio/pen/LYWPKKG) (adapted from @archmoj)

Function performance checkup:

Screenshot 2021-05-05 at 12 47 00

### CODE REVIEW

File: src/components/color/index.js [hunk 1]
```diff
@@ -116,7 +116,7 @@ color.clean = function(container) {
             if(!Array.isArray(el0) && el0 && typeof el0 === 'object') {
                 for(j = 0; j < val.length; j++) color.clean(val[j]);
             }
-        } else if(val && typeof val === 'object') color.clean(val);
+        } else if(val && typeof val === 'object' && !ArrayBuffer.isView(val)) color.clean(val);
```
Comment (archmoj) [inline src/components/color/index.js:119]: Let's only allow typed arrays here by using the `isTypedArray` function.
https://github.com/plotly/plotly.js/blob/65f739ddb75cac1e809d8a57098c52833dafd151/src/lib/array.js#L8-L11
To use that you could require it like this:
```js
var isTypedArray = require('../../lib/array').isTypedArray;
```

**Comment by archmoj** [comment]
Thanks very much for the PR.
LGTM.

**Review by alexcjohnson — APPROVED** [review]
💃

**Comment by nicolaskruchten** [comment]
Does this PR fully resolve #4401 i.e. the part about line-decimation or just the part about Arrays?

**Comment by archmoj** [comment]
> Does this PR fully resolve #4401 i.e. the part about line-decimation or just the part about Arrays?

No. I reopened the issue.

**Pull request merged**